migration/migrator.go

```diff
@@ -0,0 +1,347 @@
+package migration
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/eth/tracers"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/trie"
+	"github.com/ethereum/go-ethereum/trie/trienode"
+	"github.com/ethereum/go-ethereum/trie/zk"
+)
+
+var (
+	tracerType = "prestateTracer"
+	// BedrockTransitionBlockExtraData represents the extradata
+	// set in the very first bedrock block. This value must be
+	// less than 32 bytes long or it will create an invalid block.
+	BedrockTransitionBlockExtraData = []byte("BEDROCK")
+)
+
+type ethBackend interface {
+	ChainDb() ethdb.Database
+	BlockChain() *core.BlockChain
+}
+
+type StateMigrator struct {
+	backend       ethBackend
+	db            ethdb.Database
+	zkdb          *trie.Database
+	mptdb         *trie.Database
+	allocPreimage map[common.Hash][]byte
+	tracersAPI    *tracers.API
+	traceCfg      *tracers.TraceConfig
+	migratedRef   *core.MigratedRef
+
+	stopCh chan struct{}
+}
+
+func NewStateMigrator(backend ethBackend, tracersAPI *tracers.API) *StateMigrator {
+	db := backend.ChainDb()
+
+	allocPreimage, err := zkPreimageWithAlloc(db)
+	if err != nil {
+		log.Crit("Failed to read genesis alloc", "err", err)
+	}
+	return &StateMigrator{
+		backend: backend,
+		db:      db,
+		zkdb: trie.NewDatabase(db, &trie.Config{
+			Preimages:   true,
+			Zktrie:      true,
+			KromaZKTrie: backend.BlockChain().TrieDB().IsKromaZK(),
+		}),
+		mptdb:         trie.NewDatabase(db, &trie.Config{Preimages: true}),
+		allocPreimage: allocPreimage,
+		tracersAPI:    tracersAPI,
+		traceCfg: &tracers.TraceConfig{
+			Tracer:       &tracerType,
+			TracerConfig: json.RawMessage(`{"diffMode": true}`),
+		},
+		migratedRef: core.NewMigratedRef(db),
+		stopCh:      make(chan struct{}),
+	}
+}
+
+func (m *StateMigrator) Start() {
+	log.Info("Start state migrator to migrate ZKT to MPT")
+	go func() {
```

Comment:
How about just remove this line and just call this function like `go m.Start()`?

Reply:
This work should always be run as a goroutine, which is why I have it inside the `Start()` to force a goroutine.